common/src/main/java/com/wynntils/models/players/FriendsModel.java

```diff
@@ -0,0 +1,195 @@
+        }
+
+        if (tryParseFriendMessages(coded)) {
+            return;
+        }
+
+        if (expectingFriendMessage) {
+            if (tryParseFriendList(unformatted) || tryParseNoFriendList(coded)) {
+                event.setCanceled(true);
+                expectingFriendMessage = false;
+                return;
+            }
+
+            // Skip first message of two, but still expect more messages
+            if (FRIEND_LIST_FAIL_1.matcher(coded).matches()) {
+                event.setCanceled(true);
+                return;
+            }
+        }
+    }
+
+    private boolean tryParseNoFriendList(String coded) {
+        if (FRIEND_LIST_FAIL_2.matcher(coded).matches()) {
+            WynntilsMod.info("Player has no friends!");
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean tryParseFriendMessages(String coded) {
+        Matcher matcher = FRIEND_REMOVE_MESSAGE_PATTERN.matcher(coded);
+        if (matcher.matches()) {
+            String player = matcher.group(1);
+
+            WynntilsMod.info("Player has removed friend: " + player);
+
+            friends.remove(player);
+            WynntilsMod.postEvent(
+                    new RelationsUpdateEvent.FriendList(Set.of(player), RelationsUpdateEvent.ChangeType.REMOVE));
+            return true;
+        }
+
+        matcher = FRIEND_ADD_MESSAGE_PATTERN.matcher(coded);
+        if (matcher.matches()) {
+            String player = matcher.group(1);
+
+            WynntilsMod.info("Player has added friend: " + player);
+
+            friends.add(player);
+            WynntilsMod.postEvent(
+                    new RelationsUpdateEvent.FriendList(Set.of(player), RelationsUpdateEvent.ChangeType.ADD));
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean tryParseFriendList(String unformatted) {
+        Matcher matcher = FRIEND_LIST.matcher(unformatted);
+        if (!matcher.matches()) return false;
+
+        String[] friendList = matcher.group(1).split(", ");
+
+        friends = Arrays.stream(friendList).collect(Collectors.toSet());
+        WynntilsMod.postEvent(new RelationsUpdateEvent.FriendList(friends, RelationsUpdateEvent.ChangeType.RELOAD));
+
+        WynntilsMod.info("Successfully updated friend list, user has " + friendList.length + " friends.");
+        return true;
+    }
+
+    private void resetFriendData() {
+        friends = new HashSet<>();
+
+        WynntilsMod.postEvent(new RelationsUpdateEvent.FriendList(friends, RelationsUpdateEvent.ChangeType.RELOAD));
+    }
+
+    public void requestFriendData() {
+        if (McUtils.player() == null) return;
+
+        expectingFriendMessage = true;
+        McUtils.sendCommand("friend list");
+        WynntilsMod.info("Requested friend list from Wynncraft.");
+    }
+
+    public Set<String> getFriends() {
+        return friends;
+    }
+
+    @SubscribeEvent
+    public void onFriendJoin(FriendConnectionEvent.Join e) {
```

Comment:
This does not seem right, either. Now the model has hard-coded knowledge about the screen. Let me think a bit and see if I can suggest a better solution.